bindings/python/pydeck/pydeck/data_utils/viewport_helpers.py

```diff
@@ -0,0 +1,164 @@
+
+
+def euclidean(y, y1):
+    """Euclidean distance in n-dimensions
+
+    Parameters
+    ----------
+    y : `tuple` of `float`
+        A point in n-dimensions
+    y1 : `tuple` of `float`
+        A point in n-dimensions
+
+    Examples
+    --------
+    >>> EPSILON = 0.001
+    >>> euclidean((3, 6, 5), (7, -5, 1)) - 12.369 < EPSILON
+    True
+    """
+    if not len(y) == len(y1):
+        raise Exception('Input coordinates must be of the same length')
+    return math.sqrt(sum([_squared_diff(x, x0) for x, x0 in zip(y, y1)]))
+
+
+def geometric_mean(points):
+    """Gets centroid in a series of points
+
+    Parameters
+    ----------
+    points : :obj:`list` of :obj:`list` of :obj:`float`
+        List of (x, y) coordinates
+
+    Returns
+    -------
+        tuple : The centroid of a list of points
+    """
+    avg_x = sum([float(p[0]) for p in points]) / len(points)
+    avg_y = sum([float(p[1]) for p in points]) / len(points)
+    return (avg_x, avg_y)
+
+
+def get_bbox(points):
+    """Get the bounding box around the data,
+
+    Parameters
+    ----------
+    points : :obj:`list` of :obj:`list` of :obj:`float`
+        List of (x, y) coordinates
+
+    Returns
+    -------
+        dict: Dictionary containing the top left and bottom right points of a bounding box
+    """
+    xs = [p[0] for p in points]
+    ys = [p[1] for p in points]
+    max_x = max(xs)
+    max_y = max(ys)
+    min_x = min(xs)
+    min_y = min(ys)
+    return ((min_x, max_y), (max_x, min_y))
+
+
+def k_nearest_neighbors(points, center, k):
+    """Gets the k furthest points from the center
+
+    Parameters
+    ----------
+    points : :obj:`list` of :obj:`list` of :obj:`float`
+        List of (x, y) coordinates
+    center : :obj:`list` of :obj:`list` of :obj:`float`
+        Center point
+    k : int
+        Number of points
+
+    Returns
+    -------
+    list: Index of the k furthest points
+
+    Todo
+    ---
+    Currently implemently naively, needs to be more efficient
+    """
+    pts_with_distance = [(pt, euclidean(pt, center)) for pt in points]
+    sorted_pts = sorted(pts_with_distance, key=lambda x: x[1])
+    return [x[0] for x in sorted_pts][:int(k)]
+
+
+def get_n_pct(points, proportion=1):
+    """Computes the bounding box of the maximum zoom for the specified list of points
+
+    Parameters
+    ----------
+    points : :obj:`list` of :obj:`list` of :obj:`float`
+        List of (x, y) coordinates
+    proportion : float, default 1
+        Value between 0 and 1 representing the minimum proportion of data to be captured
+
+    Returns
+    -------
+        list: k nearest data points
+    """
+    if proportion == 1:
+        return points
+    # Compute the medioid of the data
+    centroid = geometric_mean(points)
+    # Retain the closest n*proportion points
+    n_to_keep = math.floor(proportion * len(points))
+    return k_nearest_neighbors(points, centroid, n_to_keep)
+
+
+def bbox_to_zoom_level(bbox):
```

Comment:
Use [this function](https://github.com/uber-common/viewport-mercator-project/blob/master/src/web-mercator-utils.js#L41) to project lngLat coordinates onto the WebMercator plane. A unit on the WebMercator plane can be considered equivalent to a pixel. The proper zoom level should be dependent on the width and height of the viewport.